### PR TITLE
Leap year checker

### LINKED ISSUE #7: Leap Year Checker
| **Field**        | **Details**                                                                                                         |
|------------------|---------------------------------------------------------------------------------------------------------------------|
| **Name**         | New Challenge                                                                                                       |
| **Description**  | Create a program that checks if a given year is a leap year or not.                                                |
| **Title**        | Leap Year Checker                                                                                                   |

### PR DESCRIPTION
**Pull request by Husayn01**
---
name: Leap year checker
about: A function that checks if a given year is a leap year
---

## Behavior
The leap year checker code determines whether a given year is a leap year based on specific rules defined by the Gregorian calendar. It accepts an integer input representing the year and checks whether the year is divisible by 4, ensuring it's a leap year unless divisible by 100, in which case it must also be divisible by 400 to qualify.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -122,5 +122,6 @@
       "source.fixAll.ruff": true,
       "source.organizeImports.ruff": true
     }
-  }
+  },
+  "cSpell.words": ["doctests", "Hussaini", "Pylint"]
```
Comment (NoorelsalamAlmakki) [inline .vscode/settings.json:126]: What does this part of the code do? And why it was added?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -55,7 +55,6 @@ about: A template PR for code review with a checklist
 - [ ] The function's name describes it's behavior
 - [ ] The function's name matches the file name
 - [ ] The function has correct type annotations
-- [ ] The function is not called in the function file
```
Comment (NoorelsalamAlmakki) [inline .github/PULL_REQUEST_TEMPLATE.md]: Why was this line deleted?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -83,7 +82,5 @@ about: A template PR for code review with a checklist
 - [ ] Variable names are clear and helpful
 - [ ] The code follows the strategy as simply as possible
 - [ ] The implementation is as simple as possible given the strategy
-- [ ] There are no commented lines of code
-- [ ] There are no `print` statements anywhere
```
Comment (NoorelsalamAlmakki) [inline .github/PULL_REQUEST_TEMPLATE.md]: The same goes for these lines.

File: solutions/leap_year_checker.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Module contents:
+    - is_leap_year: A function that checks if a given year is a leap year.
+
+Created on 2024-12-31
+Author: Hussaini Ahmed
+"""
+
+
+def is_leap_year(year: int) -> bool:
```
Comment (NoorelsalamAlmakki) [inline solutions/leap_year_checker.py]: The name of the function doesn't match the file name.

File: solutions/leap_year_checker.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    Parameters:
+        year (int): The year to check. Must be a positive integer.
+
+    Returns:
+        bool: True if the year is a leap year, False otherwise.
+
+    Raises:
+        TypeError: If the input is not an integer.
+        ValueError: If the year is not a positive integer.
```
Comment (NoorelsalamAlmakki) [inline solutions/leap_year_checker.py:29]: Good job! The docstring is clear and thorough.

File: solutions/leap_year_checker.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        True
+        >>> is_leap_year(2023)
+        False
+        >>> is_leap_year(0)
+        False
+    """
+    if not isinstance(year, int):
+        raise TypeError("The year must be an integer.")
+    if year <= 0:
+        raise ValueError("The year must be a positive integer.")
```
Comment (NoorelsalamAlmakki) [inline solutions/leap_year_checker.py:46]: It would be better to omit using logic in assertions. Additionally, you can use 'assert'.
```suggestion
   assert isinstance(year, int), "The year must be an integer."
   assert year > 0, "The year must be a positive integer" 
```

File: solutions/tests/test_leap_year_checker.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class TestIsLeapYear(unittest.TestCase):
+    """Test cases for the is_leap_year function."""
+
+    def test_valid_leap_years(self):
+        """Test known leap years."""
+        self.assertTrue(is_leap_year(2020))
+        self.assertTrue(is_leap_year(2000))
+        self.assertTrue(is_leap_year(1600))
```
Comment (NoorelsalamAlmakki) [inline solutions/tests/test_leap_year_checker.py]: Use only one assertion per test!

File: solutions/tests/test_leap_year_checker.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            is_leap_year(2020.5)
+        with self.assertRaises(ValueError):
+            is_leap_year(-2020)
+        with self.assertRaises(ValueError):
+            is_leap_year(0)
+
+    def test_boundary_cases(self):
+        """Test edge cases for leap year logic."""
+        self.assertFalse(is_leap_year(1))  # Minimum valid year
+        self.assertTrue(is_leap_year(4))  # First leap year
```
Comment (NoorelsalamAlmakki) [inline solutions/tests/test_leap_year_checker.py]: - The same remark applies to the rest of the tests, use only one assertion per test. 
- Use more tests to test boundary conditions.
- Use tests to check other types of inputs (e.g. float, string, list, etc.)

File: solutions/tests/test_leap_year_checker.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (NoorelsalamAlmakki) [inline solutions/tests/test_leap_year_checker.py:77]: I don't think it's necessary.

**Review by NoorelsalamAlmakki — CHANGES_REQUESTED** [review]
The overall code is good, but some areas need improvements and modifications.

**Comment by Husayn01** [comment]
@NoorelsalamAlmakki, I have optimized the code, as requested. The content deleted in the `pull request template ` was a mistake, however I have rectified that too. I was wondering why you unchecked these in the list though:

```
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Defensive assertions include as little logic as possible
```
I would need clarity on that. Also every other implementation is a personal preference, as I am sure they did not violate any guidelines, if that is not the case feel free to point it out. Thanks

**Comment by NoorelsalamAlmakki** [comment]
@Husayn01 As I understood, the checkboxes are for the reviewer to help them review the code. As a result, I unchecked them to help me properly address the comments. All in all, I have reviewed the optimized code and it's well-rectified, good work! The code is ready to be merged.

File: .vscode/settings.json (hunk 1 above)
Comment (Husayn01) [reply .vscode/settings.json:126]: My vscode do not recognize the terms in my docstring, so it throws an `unknown word` warning. Vs code recommended a quick fix, so I implemented it.

**Review by NoorelsalamAlmakki — APPROVED** [review]
All changes are addressed and the pull request is approved for merging.

**Comment by Husayn01** [comment]
> @Husayn01 As I understood, the checkboxes are for the reviewer to help them review the code. As a result, I unchecked them to help me properly address the comments. All in all, I have reviewed the optimized code and it's well-rectified, good work! The code is ready to be merged.

Oh, my bad... I thought I was to do that, then the reviewer reviews it. Sorry about that and thanks for the review.
